### PR TITLE
[WIP] Reset tenant_id as code

### PR DESCRIPTION
**Pull request by hallelujah**
## Issue description

When doing migration with pt-osc triggers are removed, thus new record in tables do not have the `tenant_id`, which can be problematic (data leak)

* The https://github.com/3scale/porta/blob/587505f9971dbbf829e7671c9a488243c1be2b83/lib/tasks/multitenant/triggers.rake cannot really be run on production environments.
Those requests are going to block the tables.
* Some are not compatible with all supported databases (MySQL, Postgres, Oracle 12c)
* Not really testable
* Cannot launch a reset per table/model

## Goals

* Make it database agnostic
* Add testing because it can be scary to have tenant_id set the wrong way. Possible data leak ...
* Definition in each model:
  * Why? that way it is handy to do `Account.reset_tenant_id!(:buyers)` Rather than ResetTenantId.call(Account, :buyers)
  *  if there is no model, use Arel to create the query (database agnostic)

### CODE REVIEW

File: lib/tasks/multitenant/update.rake [hunk 1]
```diff
@@ -0,0 +1,102 @@
+# __END__
```
Comment (hallelujah) [inline lib/tasks/multitenant/update.rake]: This file is not needed

File: app/lib/reset_tenant_id.rb [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+      if name
+        runner.execute(name)
+      else
+        runner.execute_all
+      end
+    end
+  end
+end
+
+__END__
```
Comment (hallelujah) [inline app/lib/reset_tenant_id.rb]: I will remove that

File: lib/tasks/multitenant/update.rake [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+MASTER_ID = Account.master.id
```
Comment (guicassolato) [inline lib/tasks/multitenant/update.rake]: Maybe `or raise`

File: app/lib/reset_tenant_id.rb [hunk 4]
```diff
@@ -0,0 +1,131 @@
+module ResetTenantId
+  extend ActiveSupport::Concern
```
Comment (guicassolato) [inline app/lib/reset_tenant_id.rb]: Why extending `ActiveSupport::Concern`?

File: app/models/account.rb [hunk 5]
```diff
@@ -594,5 +594,8 @@ def master_uniqueness
     errors.add :master, 'can be only one' if scope.exists?(master: true)
   end
 
-  protected
+  reset_tenant_id do
```
Comment (guicassolato) [inline app/models/account.rb]: Maybe we could find a name for that method that makes it clear that reset tenant_id will be performed when this is loaded, like `config_ reset_tenant_id` or `define_reset_tenant_id`. WDYT?

File: app/lib/reset_tenant_id.rb [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+    end
+  end
+
+  class Query
+    attr_reader :collection, :name, :with
+
+    def initialize(name, options={}, &block)
+      @name = name.to_sym
+      @collection = options[:collection] || @name
+      @with = options[:with] || block
```
Comment (guicassolato) [inline app/lib/reset_tenant_id.rb]: Nice

File: app/lib/reset_tenant_id.rb [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+  def reset_tenant_id!(name)
+    runner = ResetTenantId::Runner.new(self.class, ResetTenantId::Configurator.instance.definitions)
+    query = runner.find_query(name)
+    query.update.call(self) unless query
+  end
+
+  module ClassMethods
+
+    # define_reset_tenant_id do
+    #   updates(:buyers, with: :provider_account_id)
```
Comment (guicassolato) [inline app/lib/reset_tenant_id.rb]: Not sure if `with` is self-explanatory. It could also be `using` or `source`.

Comment (hallelujah) [reply app/lib/reset_tenant_id.rb]: well it should be english-ish readable.

`updates buyers source provider_account_id` is not really english
`updates buyers using provider_account_id` is OK but how is it different from
`updates buyers with provider_account_id`

Comment (guicassolato) [reply app/lib/reset_tenant_id.rb]: It's a good point. My concern because it's not clear what the `provider_account_id` is for. I mean we know this is where we get the value from to set `tenant_id`, but does `with` say that? Maybe there's nothing better. IDK.

File: app/concerns/reset_tenant_id.rb [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+
+    # :nodoc:
+    # Returns an object responding to #call(record)
+    def wrap_update(with)
+      case with
+      when NilClass
+        raise ArgumentError, 'Pass in `:with` option or a block'
+      when Symbol
+        proc do |record|
+          record.update_column :tenant_id, record.public_send(with) # rubocop:disable Rails/SkipsModelValidations
```
Comment (guicassolato) [inline app/concerns/reset_tenant_id.rb]: I think the API is confusing. When you pass a `Symbol`, it will call that symbol (`public_send`) on the `record`, so the value that the symbol "returns"(*) is the value moved to `tenant_id`. When you pass something that responds to `call`, then it's _not_ the value returned by the proc that will be used to set `tenant_id` but it becomes responsibility of the proc to set the `tenant_id`. In other words, `with` can be an alternative to passing a block (a complex operation that is expected to update the record), but it can also be the name of a method that will just return a scalar value.

I would expect `with` always to resolve to a scalar value. It can be a symbol or a proc, but in both cases the value returning value will be used with `record.update_column :tenant_id, ...`. And if you want to override `record.update_column :tenant_id, ...` then you have to pass a block and forget about `with`.

(*) Actually it's the value that _the method named equal to the symbol_ returns.
______

In a nutshell, something like:

```ruby
def wrap_update(with)
  raise ArgumentError, 'Pass in `:with` option or a block' if with.blank?

  proc do |record|
    record.update_column :tenant_id, with.is_a?(Symbol) ? record.public_send(with) : with.call
  end
end
```

Comment (hallelujah) [reply app/concerns/reset_tenant_id.rb]: Well it is the same as in rails :) I do not see that confusing.

`Symbol#to_proc`

Comment (hallelujah) [reply app/concerns/reset_tenant_id.rb]: It was intended by default it would use update_column but if you want to do something more particular then it will use the block or the with.

File: app/concerns/reset_tenant_id.rb [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+
+    attr_reader :klass, :name, :update, :collection
+
+    def initialize(klass, name, options = {}, &block)
+      @klass = klass
+      @name = name.to_sym
+      @collection = wrap_collection(options[:collection] || @name)
+      value = options[:value]
+      if value
+        options[:with] = -> (record) {
```
Comment (guicassolato) [inline app/concerns/reset_tenant_id.rb]: Maybe we could extract this to a private method.

Comment (guicassolato) [reply app/concerns/reset_tenant_id.rb]:
```ruby
def initialize(klass, name, options = {}, &block)
  # [...]

  value = options[:value]
  @options[:with] = self.method(:update_record_with_value).curry[value]

  # [...]
end

private

def update_record_with_value(value, record)
  record.update_column(:tenant_id, value.is_a?(Symbol) ? record.public_send(value) : value)
end
```

**Comment by Martouta** [comment]
imo we should create a jira issue for this and estimate it and follow the whole process we are doing now.

**Comment by Martouta** [comment]
I think this closes [THREESCALE-1899](https://issues.jboss.org/browse/THREESCALE-1899) and you should link them 😄

**Comment by Martouta** [comment]
This is related to [THREESCALE-1899](https://issues.jboss.org/browse/THREESCALE-1899). I asked Hery since now that we closed the Jira issue, if we can close this PR too. Then he said:

> You can but I want to continue it later
> I think we should do it by code because we have 3 databases to handle with many different syntax
> but feel free to close it :)
> It will be easy to cherry-pick the commits later

I also think that we should do it by code but we can always reopen it or cherry-pick when we are actively working on it or at least when we have its correspondent Jira issue in Next 😄
